### PR TITLE
Send and receive Exchange v2.1 bye message

### PR DESCRIPTION
**Pull request by MiSikora**
Task/Issue URL: https://app.asana.com/1/137249556945/project/1216103556496795/task/1217367677188794?focus=true
Tech Design URL (if applicable): https://app.asana.com/1/137249556945/project/1216103556496795/task/1217626160385689?focus=true
API Proposals URL(s) (if applicable): N/A

### Description

Adds the v2.1 `bye` message to the Exchange pairing protocol, in both directions. A device leaving the exchange now tells its peer it is gone before deleting its relay channel.

Sending:
- Every session teardown now writes a `bye` as its last message before the channel is deleted.
- The reason says why the device left: `done` when there is nothing left to do, `cancelled` when the user backed out, `error` on a local failure.
- The message is sent once, never retried, and delivery is never depended on.
- A message that is skipped or fails to send is now recorded as a `MessageNotSent` event; an undeliverable `bye` produces no session error, since the peer usually deletes its channel first.

Receiving:
- A peer's `bye` is accepted at any point of the exchange.
- Received before the recovery code was handed over, it ends the session immediately as a peer-left failure instead of leaving the device waiting for the five-minute deadline.
- A host waiting for the joiner's status report treats it as outcome unknown, not as a failure, because the joiner may well have succeeded.
- A joiner that is already signing in with the received recovery code carries on unaffected. Its final report is simply never delivered to the peer that left.

### Steps to test this PR

_Cancelling mid-exchange ends the peer's session immediately_
- [ ] Install this build on two devices, neither signed in to Sync & Backup.
- [ ] Open Sync Dev Settings on both devices.
- [ ] Tap "V2 Pairing Debug" on both devices.
- [ ] Tap "Change" next to "Protocol version" on both devices.
- [x] Select v2.1 on both devices.
- [ ] On the first device, turn off "Auto-approve confirmations".
- [ ] On the first device, tap "Start as Presenter".
- [ ] On the first device, tap "Copy linking code".
- [ ] Transfer the linking code to the second device's clipboard.
- [ ] On the second device, tap "Paste & Go".
- [ ] Wait for the "Confirm pairing (Host)" dialog on the first device.
- [ ] On the second device, tap "Cancel".
- [ ] Verify the second device's event log shows a sent `bye` with `reason=cancelled`.
- [ ] Verify the first device fails the pairing immediately, without waiting for the five-minute deadline.
- [ ] Verify the first device's event log shows the received `bye` as PeerLeft.

_Successful v2.1 pairing says bye with reason done_
- [ ] On the first device, turn "Auto-approve confirmations" back on.
- [ ] On the first device, tap "Start as Presenter".
- [ ] On the first device, tap "Copy linking code".
- [ ] Transfer the linking code to the second device's clipboard.
- [ ] On the second device, tap "Paste & Go".
- [ ] Verify pairing completes on both devices.
- [ ] Verify each device's event log shows a sent `bye` with `reason=done`, or a "Not sent bye" row when the peer's channel was already gone.

### UI changes
N/A

---

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

---

> [!NOTE]
> **Medium Risk**
> Changes core v2 pairing teardown, state-machine transitions, and failure telemetry mapping; behavior differs by negotiated protocol version and pairing phase, with broad test coverage but real multi-device edge cases remain.
> 
> **Overview**
> Adds the v2.1 **`bye`** wire message so a device can signal it is leaving the exchange before deleting its relay channel, instead of leaving the peer on a long timeout.
> 
> **Outbound:** Session teardown now sends a best-effort `bye` (once, before channel DELETE) with reason **`done`**, **`cancelled`**, or **`error`** depending on how the session ended. Peers on negotiated v2.0 skip sending `bye` (too high protocol). Failed or skipped sends emit **`MessageNotSent`**; a failed `bye` does not raise **`SessionError`**.
> 
> **Inbound:** The state machine accepts `bye` in every state. Early in pairing it ends the session as **peer left** (`RejectReason.PeerLeft`); on the host waiting for join status it moves to **Host.Unknown**; while the joiner is logging in it **does not** stop in-flight work. **`RealSyncCodeDispatcher`** maps peer-driven aborts to **`peer_left(...)`** failures with **`PAIRING_REJECTED`** or **`PAIRING_FAILED`** when the bye reason is error.
> 
> Debug UI labels for reject/not-sent reasons are updated for the new events.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0aba25c8261935209c0c1d334ef831993e5d6def. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/exchange/v2/ExchangeV2StateMachine.kt [hunk 1]
```diff
@@ -204,10 +211,38 @@ internal class RealExchangeV2StateMachine(
...
+        // No report is coming, but the peer may well have joined; never a failure.
+        ExchangeV2State.Host.AwaitingStatus,
+        ExchangeV2State.Host.Unknown,
+        -> accept(state, ExchangeV2State.Host.Unknown, msg)
+        // Past the point of no return: the peer going away is not a reason to abandon a login,
+        // upgrade or sync that is already running. The later recovery_code_done just won't land.
+        ExchangeV2State.Joiner.Joining -> accept(state, state, msg)
+        // Before the recovery code is released there is nobody left to pair with. In an already
+        // terminal state abort() keeps the state and only records the message.
+        else -> abort(state, msg, RejectReason.PeerLeft)
```
[the remark this answers is not in the document]
Comment (MiSikora) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/exchange/v2/ExchangeV2StateMachine.kt:243]: This is true, but I prefer to leave it for now. It's not an issue right now because the feature flag is off and `Bye` is never dispatched.

I want to clean up the debug screen first, as it duplicates a lot of production logic, which makes it difficult to manage.

I'll address all of that in: https://app.asana.com/1/137249556945/task/1218094251260965

**Review by CDRussell — APPROVED** [review]
